core/src/main/java/info/rmapproject/core/idservice/HttpArkIdService.java

```diff
@@ -0,0 +1,395 @@
+     * @return boolean
+     * @throws Exception the exception
+     */
+    private boolean isValidId(String id) throws Exception {
+        boolean isValid = true;
+        if (idRegex!=null && idRegex.length()>0){
+            isValid = id.matches(idRegex);
+        }
+        if (isValid && idLength>0) {
+            isValid = (id.length()==idLength);
+        }
+        return isValid;
+    }
+
+    /* (non-Javadoc)
+     * @see info.rmapproject.core.idservice.IdService#createId()
+     */
+    public URI createId() throws Exception {
+        try {
+            return new URI(getEzid());
+        } catch (Exception e) {
+            throw new Exception("Failed to create a new ID.", e);
+        }
+    }
+
+    private synchronized String getEzid() throws Exception {
+        String id;
+        MVStore mvs = MVStore.open(idStoreFile);
+        ezids = mvs.openMap(DATA);
+        //if we are out of IDs, let's get some more from the EZID service endpoint
+        if(ezids.size() < 1){
+           if(!replenishingCache) {
+               getMoreEzids();
+           }
+        }
+
+        //let's make sure we have one to hand out. we can wait just a bit to let the previous
+        //call fill up the cache a bit
+        boolean shouldRetry;
+        int retryCounter = 0;
+        do{
+            retryCounter++;
+            shouldRetry = (retryCounter < maxRetryAttempts && ezids.size() == 0);
+			//WAIT FOR 5 SECS BEFORE RE-TRYING TO OVERCOME TEMPORARY NETWORK FAILURES
+			//OR THE EZID SERVER BEING BUSY SERVICING ANOTHER REQUEST.
+			if(shouldRetry) {
+                try {
+                    wait(RETRY_WAIT_TIME);
+                } catch (InterruptedException ie) {
+                    log.error("Wait interrupted in retry loop", ie);
+                }
+            }
+        } while (shouldRetry);
+
+
+        //find the first key for which we have an EZID
+        Integer i = 1;
+        while(!ezids.containsKey(i) && i <= maxStoreSize) {
+            i++;
+        }
+
+        //failsafe - if we still don't have any, we throw an exception
+        if(i>maxStoreSize){
+            throw new RuntimeException("EZID cache is empty, unable to provide one");
+        }
+
+        //we take this id off of our cache
+        id=ezids.remove(i);
+
+        mvs.close();
+        return id;
+    }
+
+    private void getMoreEzids() {
+        replenishingCache=true;
+        EZIDClient client = new EZIDClient(serviceUrl);
+        boolean shouldRetry;
+        int retryCounter = 0;
+		do {
+            retryCounter++;
+            log.debug("Minting ids from " + serviceUrl);
+            try {
+                client.login(userName, userPassword);
+                for (Integer i = 1; !ezids.keyList().contains(i) && i <= maxStoreSize; i++) {//first test condition allows us to refresh a partially populated cache in the future
+                    String id = client.mintIdentifier(idPrefix,null);
+                    if(isValidId(id)) {
```

Comment:
Is it necessary to test the "validity" of the id returned from the service?  If so, no problem.  If it isn't necessary, then maybe we can just eliminate code associated with testing a valid url.